prog/prog.go

```diff
@@ -15,113 +15,100 @@ type Prog struct {
 
 type Call struct {
 	Meta *sys.Call
-	Args []*Arg
-	Ret  *Arg
-}
-
-type Arg struct {
-	Type         sys.Type
-	Kind         ArgKind
-	Val          uintptr       // value of ArgConst
-	AddrPage     uintptr       // page index for ArgPointer address, page count for ArgPageSize
-	AddrOffset   int           // page offset for ArgPointer address
-	AddrPagesNum uintptr       // number of available pages for ArgPointer
-	Data         []byte        // data of ArgData
-	Inner        []*Arg        // subargs of ArgGroup
-	Res          *Arg          // target of ArgResult, pointee for ArgPointer
-	Uses         map[*Arg]bool // this arg is used by those ArgResult args
-	OpDiv        uintptr       // divide result for ArgResult (executed before OpAdd)
-	OpAdd        uintptr       // add to result for ArgResult
-
-	// ArgUnion/UnionType
-	Option     *Arg
-	OptionType sys.Type
+	Args []Arg
+	Ret  Arg
 }
 
-type ArgKind int
+type Arg interface {
+	Type() sys.Type
+	Size() uintptr
+	Value(pid int) uintptr // returns value taking endianness and executor pid into consideration
+	Uses() *map[Arg]bool
+}
 
-const (
-	ArgConst ArgKind = iota
-	ArgResult
-	ArgPointer  // even if these are always constant (for reproducibility), we use a separate type because they are represented in an abstract (base+page+offset) form
-	ArgPageSize // same as ArgPointer but base is not added, so it represents "lengths" in pages
-	ArgData
-	ArgGroup // logical group of args (struct or array)
-	ArgUnion
-	ArgReturn // fake value denoting syscall return value
-)
+type ArgCommon struct {
+	typ  sys.Type
+	uses map[Arg]bool // this arg is used by ArgResult args
+}
 
-// Returns inner arg for PtrType args
-func (a *Arg) InnerArg() *Arg {
-	switch typ := a.Type.(type) {
-	case *sys.PtrType:
-		if a.Res == nil {
-			if !typ.Optional() {
-				panic(fmt.Sprintf("non-optional pointer is nil\narg: %+v\ntype: %+v", a, typ))
-			}
-			return nil
-		} else {
-			return a.Res.InnerArg()
-		}
-	default:
-		return a
-	}
+func (arg *ArgCommon) Type() sys.Type {
+	return arg.typ
 }
 
-func encodeValue(value, size uintptr, bigEndian bool) uintptr {
-	if !bigEndian {
-		return value
-	}
-	switch size {
-	case 2:
-		return uintptr(swap16(uint16(value)))
-	case 4:
-		return uintptr(swap32(uint32(value)))
-	case 8:
-		return uintptr(swap64(uint64(value)))
-	default:
-		panic(fmt.Sprintf("bad size %v for value %v", size, value))
-	}
+func (arg *ArgCommon) Size() uintptr {
+	return arg.typ.Size()
+}
+
+func (arg *ArgCommon) Value(pid int) uintptr {
+	return 0
 }
 
-// Returns value taking endianness into consideration.
-func (a *Arg) Value(pid int) uintptr {
-	switch typ := a.Type.(type) {
+func (arg *ArgCommon) Uses() *map[Arg]bool {
+	return &arg.uses
```

Comment:
This is fishy. Please try to remove this method and uses from common part. Uses are only relevant for ResultArg and ReturnArg.